File: android/app/src/main/java/com/shoply/app/PushTokenStore.kt
package com.shoply.app

import android.content.Context
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions

object PushTokenStore {
    private const val PREFS = "shoply"
    private const val KEY = "fcmToken"
    private val db = FirebaseFirestore.getInstance()

    fun updateToken(context: Context, token: String) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putString(KEY, token)
            .apply()
        saveIfPossible(context)
    }

    fun sync(context: Context) {
        saveIfPossible(context)
    }

    private fun saveIfPossible(context: Context) {
        val token = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(KEY, null)
            ?: return
        val uid = FirebaseAuth.getInstance().currentUser?.uid ?: return
        val data = hashMapOf(
            "token" to token,
            "platform" to "android",
            "updatedAt" to FieldValue.serverTimestamp()
        )
        db.collection("users").document(uid).collection("tokens").document(token)
            .set(data, SetOptions.merge())
    }
}

File: ios/Shoply/Services/PushTokenStore.swift
import FirebaseAuth
import FirebaseFirestore
import Foundation

final class PushTokenStore {
    static let shared = PushTokenStore()
    private let db = Firestore.firestore()
    private let tokenKey = "fcmToken"

    private init() {}

    func updateToken(_ token: String) {
        UserDefaults.standard.set(token, forKey: tokenKey)
        saveIfPossible()
    }

    func syncIfNeeded() {
        saveIfPossible()
    }

    private func saveIfPossible() {
        guard let token = UserDefaults.standard.string(forKey: tokenKey),
              let userId = Auth.auth().currentUser?.uid else { return }
        let data: [String: Any] = [
            "token": token,
            "platform": "ios",
            "updatedAt": FieldValue.serverTimestamp()
        ]
        db.collection("users").document(userId).collection("tokens").document(token).setData(data, merge: true)
    }
}
